docs: guide quick start app creation using the slack cli (#2535)
slackapi/bolt-js: docs/content/building-an-app.mdx

---
title: Building an app with Bolt for JavaScript
sidebar_label: Building an app
---

This guide is meant to walk you through getting up and running with a Slack app using Bolt for JavaScript. Along the way, we’ll create a new Slack app, set up your local environment, and develop an app that listens and responds to messages from a Slack workspace.

> When you’re finished, you’ll have created the [Getting Started app](https://github.com/slackapi/bolt-js-getting-started-app) to run, modify, and make your own. ⚡️

Reviewer: ```suggestion
When you’re finished, you’ll have created the [Getting started app](https://github.com/slackapi/bolt-js-getting-started-app) to run, modify, and make your own. ⚡️
```

---

## Create an app {#create-an-app}
First thing's first: before you start developing with Bolt, you'll want to [create a Slack app](https://api.slack.com/apps/new).

> :::tip

Author: ```suggestion
:::tip[A place to test and learn]
```

If you don’t already have a project, let’s create a new one. Create an empty directory and initialize a new project:

```shell
mkdir first-bolt-app
cd first-bolt-app
npm init
```

You’ll be prompted with a series of questions to describe your new project (you can accept the defaults by hitting <kbd>Enter</kbd> on each prompt if you aren’t picky). After you’re done, you’ll have a new `package.json` file in your directory.

Before we install the Bolt for JavaScript package to your new project, let's save the **bot token** and **Signing Secret** that were generated when you configured your app.

1. **Copy your Signing Secret from the Basic Information page** and then store it in a new environment variable. The following example works on Linux and macOS; but [similar commands are available on Windows](https://superuser.com/questions/212150/how-to-set-env-variable-in-windows-cmd-line/212153#212153).

```shell
export SLACK_SIGNING_SECRET=<your-signing-secret>
```

2. **Copy your bot (xoxb) token from the OAuth & Permissions page** and store it in another environment variable.
```shell
export SLACK_BOT_TOKEN=xoxb-<your-bot-token>
```

> :::warning

Reviewer: ```suggestion
:::warning[Keep it secret. Keep it safe]
```

Author: A similar callout is made in an above callout but I am hoping we can revisit parts of this guide altogether soon since these seem to repeat the same ideas multiple times it feels 🔏

</TabItem>
<TabItem value="http" label="HTTP">

1. Go back to your app configuration page (click on the app from your [app settings page](https://api.slack.com/apps)). Click **Event Subscriptions** on the left sidebar. Toggle the switch labeled **Enable Events**.

2. Add your Request URL. Slack will send HTTP POST requests corresponding to events to this [Request URL](https://docs.slack.dev/apis/events-api/#subscribing) endpoint. Bolt uses the `/slack/events` path to listen to all incoming requests (whether shortcuts, events, or interactivity payloads). When configuring your Request URL within your app configuration, you'll append `/slack/events`, e.g. `https://<your-domain>/slack/events`. 💡

:::info

For local development, you can use a proxy service like [ngrok](https://ngrok.com/) to create a public URL and tunnel requests to your development environment. Refer to [ngrok's getting started guide](https://ngrok.com/docs#getting-started-expose) on how to create this tunnel.

:::

</TabItem>
</Tabs>

Scroll down to **Subscribe to Bot Events**. There are four events related to messages:

> - [`message.channels`](https://docs.slack.dev/reference/events/message.channels) listens for messages in public channels that your app is added to
> - [`message.groups`](https://docs.slack.dev/reference/events/message.groups) listens for messages in 🔒 private channels that your app is added to
> - [`message.im`](https://docs.slack.dev/reference/events/message.im) listens for messages in your app's DMs with users
> - [`message.mpim`](https://docs.slack.dev/reference/events/message.mpim) listens for messages in multi-person DMs that your app is added to

Reviewer: ```suggestion
- [`message.channels`](https://docs.slack.dev/reference/events/message.channels) listens for messages in public channels that your app is added to.
- [`message.groups`](https://docs.slack.dev/reference/events/message.groups) listens for messages in 🔒 private channels that your app is added to.
- [`message.im`](https://docs.slack.dev/reference/events/message.im) listens for messages in your app's DMs with users.
- [`message.mpim`](https://docs.slack.dev/reference/events/message.mpim) listens for messages in multi-person DMs that your app is added to.
```

<Tabs groupId="socket-or-http">
<TabItem value="socket-mode" label="Socket Mode">

With Socket Mode on, basic interactivity is enabled for us by default, so no further action here is needed

</TabItem>
<TabItem value="http" label="HTTP">

Similar to events, you'll need to specify a Request URL for Slack to send the action (such as *user clicked a button*).

By default, Bolt uses the same endpoint for interactive components that it uses for events, so use the same request URL as above (in the example, it was `https://8e8ec2d7.ngrok.io/slack/events`). Press the **Save Changes** button in the lower right hand corner, and that's it. Your app is set up to handle interactivity!

</TabItem>
</Tabs>

When interactivity is enabled, interactions with shortcuts, modals, or interactive components (such as buttons, select menus, and datepickers) will be sent to your app as events.

Now, let’s go back to your app’s code and add logic to handle those events:
- First, we'll send a message that contains an interactive component (in this case a button).

> - Next, we'll listen for the action of a user clicking the button before responding

Reviewer: ```suggestion
- Next, we'll listen for the action of a user clicking the button before responding.
```

  await say({
    blocks: [
      {
        "type": "section",
        "text": {
          "type": "mrkdwn",
          "text": `Hey there <@${message.user}>!`
        },
        "accessory": {
          "type": "button",
          "text": {
            "type": "plain_text",
            "text": "Click Me"
          },
          "action_id": "button_click"
        }
      }
    ],
    text: `Hey there <@${message.user}>!`
  });
});

(async () => {
  // Start your app
  await app.start(process.env.PORT || 3000);

  app.logger.info('⚡️ Bolt app is running!');
})();
```
  </TabItem>
</Tabs>

The value inside of `say()` is now an object that contains an array of `blocks`. Blocks are the building components of a Slack message and can range from text to images to datepickers. In this case, your app will respond with a section block that includes a button as an accessory. Since we're using `blocks`, the `text` is a fallback for notifications and accessibility.

You'll notice in the button `accessory` object, there is an `action_id`. This will act as a unique identifier for the button so your app can specify what action it wants to respond to.

> :::tip

Reviewer: ```suggestion
:::tip[Using Block Kit Builder]
```

> The [Block Kit Builder](https://app.slack.com/block-kit-builder) is a simple way to prototype your interactive messages. The builder lets you (or anyone on your team) mockup messages and generates the corresponding JSON that you can paste directly in your app.

Reviewer: ```suggestion
The [Block Kit Builder](https://app.slack.com/block-kit-builder) is a simple way to prototype your interactive messages. The builder lets you (or anyone on your team) mock up messages and generates the corresponding JSON that you can paste directly in your app.
```

like setup / set up noun/verb distinction

});

(async () => {
  // Start your app
  await app.start(process.env.PORT || 3000);

  app.logger.info('⚡️ Bolt app is running!');
})();
```

  </TabItem>
</Tabs>

You can see that we used `app.action()` to listen for the `action_id` that we named `button_click`. If you restart your app and click the button, you'll see a new message from your app that says you clicked the button.

---

## Next steps {#next-steps}
You just built your first [Bolt for JavaScript app](https://github.com/slackapi/bolt-js-getting-started-app)! 🎉

Now that you have a basic app up and running, you can start exploring how to make your Bolt app stand out. Here are some ideas about what to explore next:

* Read through the concepts pages to learn about the different methods and features your Bolt app has access to.

* Explore the different events your bot can listen to with the [`events()`](/concepts/event-listening) method. All of the events are listed [on the API site](https://docs.slack.dev/reference/events).

* Bolt allows you to [call Web API methods](/concepts/web-api) with the client attached to your app. There are [over 200 methods](https://docs.slack.dev/reference/methods) on our API site.

> * Learn more about the different token types [on our API site](https://docs.slack.dev/authentication/tokens). Your app may need different tokens depending on the actions you want it to perform.

Reviewer: ```suggestion
* Learn more about the different token types [on the API docs site](https://docs.slack.dev/authentication/tokens). Your app may need different tokens depending on the actions you want it to perform.
```
